locator/topology.cc

```diff
@@ -337,20 +346,38 @@ void topology::index_node(node* node, const endpoint_dc_rack& dr) {
             loc.rack = _cfg.local_dc_rack.rack;
         }
     }
-    const auto& endpoint = node->endpoint();
-    // FIXME: provide storing exception safety guarantees
-    auto& dc = loc.dc;
-    auto& rack = loc.rack;
-    _dc_nodes[dc].emplace(node);
-    _dc_rack_nodes[dc][rack].emplace(node);
-    _dc_endpoints[dc].insert(endpoint);
-    _dc_racks[dc][rack].insert(endpoint);
-    _datacenters.insert(dc);
 
-    node->_dc_rack = std::move(loc);
+    datacenter* dc = nullptr;
+    if (auto dc_it = _datacenters.find(loc.dc); dc_it != _datacenters.end()) {
+        dc = const_cast<datacenter*>(dc_it->second.get());
+    } else {
+        auto new_dc = std::make_unique<datacenter>(loc.dc);
+        dc = new_dc.get();
+        _datacenters.emplace(new_dc->name(), std::move(new_dc));
+    }
 
-    if (node->is_this_node()) {
-        _this_node = node;
+    rack* rack = nullptr;
+    if (auto rack_it = dc->racks().find(loc.rack); rack_it != dc->racks().end()) {
+        rack = const_cast<locator::rack*>(rack_it->second.get());
+    } else {
+        auto new_rack = std::make_unique<locator::rack>(loc.rack);
+        rack = new_rack.get();
+        dc->racks().emplace(new_rack->name(), std::move(new_rack));
+    }
+
+    // Provide strong exception safety guarantees.
+    // No exceptions are allowed after insertion of node succeeded
+    // Note: it is okay to leave behind empty dc/rack since they are ignored
```

Comment:
Is this path called outside of mutate_token_metadata()? If not, we don't need strong exception guarantees here. If it is, then it shouldn't be because users assume topology held by token_metadata_ptr is immutable.

Reply:
Even if it is so, I think it's healthy to guarantee those here since it doesn't make the code much more complicated.

Reply:
Is it really strongly exception safe though? Before this line we modify members like _nodes_by_endpoint and _datacenters and don't roll back on exception.

Reply:
You're right, we do rely on the token_metadata/topology immutability so we don't require add_node to be exception safe. I removed that code from the latest version